Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -373,372 +374,401 @@ namespace MaterialCanvas
+                return nodeB->HasInputConnectionFromNode(nodeA);
             });
 
-        return nodes;
+        return sortedNodes;
     }
 
-    bool MaterialCanvasDocument::CompileGraph() const
+    void MaterialCanvasDocument::ReplaceStringsInContainer(
+        const AZStd::string& findText, const AZStd::string& replaceText, AZStd::vector<AZStd::string>& container) const
     {
-        m_generatedFiles.clear();
-
-        // All slots and nodes will be visited to collect all of the unique include and template paths.
-        AZStd::set<AZStd::string> includePaths;
-        AZStd::set<AZStd::string> templatePaths;
-
-        // Build a unique set of settings found on a node or slot configuration.
-        auto collectSettingsAsSet = [](const AtomToolsFramework::DynamicNodeSettingsMap& settings,
-                                       const AZStd::string& settingName,
-                                       AZStd::set<AZStd::string>& container)
+        for (auto& sourceText : container)
         {
-            if (auto settingsItr = settings.find(settingName); settingsItr != settings.end())
-            {
-                container.insert(settingsItr->second.begin(), settingsItr->second.end());
-            }
-        };
+            AZ::StringFunc::Replace(sourceText, findText.c_str(), replaceText.c_str());
+        }
+    }
 
-        // There's probably no reason to distinguish between function and class definitions. This could really be any globally defined
-        // function, class, struct, define.
-        AZStd::vector<AZStd::string> classDefinitions;
-        AZStd::vector<AZStd::string> functionDefinitions;
+    AZStd::string MaterialCanvasDocument::ConvertSlotTypeToAZSL(const AZStd::string& slotTypeName) const
+    {
+        if (AZ::StringFunc::Equal(slotTypeName, "color"))
+        {
+            return "float4";
+        }
 
-        // This is a complete list of all the instructions collected from settings found in each node.
-        // The set of instructions will need to be rebuilt for every instance of a generated code block based on the context and outputs
-        // expected from that block. This will be updated in another pass so that instruction can be built by filtering the set of nodes
-        // down to only those that will be used to generate the outputs listed in each block.
-        AZStd::vector<AZStd::string> instructions;
+        return slotTypeName;
+    }
 
-        // Build an accumulated list of settings found on a node or slot configuration.
-        auto collectSettingsAsVec = [](const AtomToolsFramework::DynamicNodeSettingsMap& settings,
-                                       const AZStd::string& settingName,
-                                       AZStd::vector<AZStd::string>& container)
+    AZStd::string MaterialCanvasDocument::ConvertSlotValueToAZSL(const AZStd::any& slotValue) const
+    {
+        if (auto v = AZStd::any_cast<const AZ::Color>(&slotValue))
```

Comment:
Looking at this, I am sort of wishing we just tagged the type explicitly instead of needing to rely on an RTTI cast. Is there no tagged-union (or dare I say it, variant) support for graph model?

Reply:
Yes, when we register data types with graph model, they are registered with a type name. We can piggyback off of that. This portion is temporary. It will be replaced with a more scalable mechanism to register conversion functions with the data type since we are going to have to support several others. Will need to explore options for matrix and other types not conventionally displayed on nodes. Perhaps those will just show up in the inspector.